### ISSUE TITLE
Overzealous regex error

### ISSUE THREAD

**Issue by librasteve** [rakudo/rakudo#5737]
## The Problem

None regex code is causing a regex syntax error.

## Expected Behavior

```
use Physics::Constants; 
use Physics::Measure :ALL;

my \λ = 2.5nm; 
my \ν = c / λ;  

say "Frequency of photon (ν) is " ~ν.norm;          #119.92 petahertz 
```

## Actual Behavior

Getting error:

```
Unable to parse regex; couldn't find final '/'
at /Users/xxx/xxx/synopsis-constants.raku:4
------> my \ν = c / λ;⏏  
```

## Steps to Reproduce
- `zef install Physics::Measure, Physics::Constants`
- `synopsis-constants.raku`

## Environment
* macOs sequoia 15.11 (M1)
* Welcome to Rakudo™ v2024.10.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2024.10.

**Comment by jubilatious1** [comment]
Doesn't like `;` semicolon:

```raku
 %  raku -e 'my $λ = 2.5 ; my $ν = c / $λ ; say "Frequency of photon (ν) is " ~$ν;    #119.92 petahertz ;'

===SORRY!===
Unrecognized regex metacharacter ; (must be quoted to match literally)
at -e:1
------> my $λ = 2.5 ; my $ν = c / $λ⏏ ; say "Frequency of photon (ν) is " ~$ν
Unable to parse regex; couldn't find final '/'
at -e:1
------> my $λ = 2.5 ; my $ν = c / $λ ;⏏ say "Frequency of photon (ν) is " ~$ν;
```
____
But maybe there's a problem with declaration/instantiation of  `c` ?

```raku
% raku -e 'my $λ = 2.5 ; my $c = 3e9; my $ν = $c / $λ ; say "Frequency of photon (ν) is " ~$ν;    #119.92 petahertz ;'

Frequency of photon (ν) is 1200000000
```

**Comment by librasteve** [comment]
@jubilatious1 - thank you for your comment

of course my MRE is incomplete in my desire to make a minimal example, I have edited to include `Physics::Constants`

**Comment by ab5tract** [comment]
I wonder if wrapping the measure object and / or installing as &term:<c> might have any impact?

**Comment by jubilatious1** [comment]
@librasteve so you run a regex to capture/remove the units `nm` from ` \λ`, then work from there? 
As you've probably realized/implemented, you'll need to capture/clean-up any `/` forward-slashes, lest they get recognized as regex-matchers: 
```raku
S[old] = "new"
```
...might be helpful: https://docs.raku.org/language/regexes#Lexical_conventions_0

**Comment by librasteve** [comment]
I plan to evolve Physics::Measure to a RakuAST slang - which will in the meantime I think the workaround I described (remove ws around the `/` operator) is adequate for what I need. 

As mentioned above, I am leaving this issue open as I consider it to be a small but non zero bug in the raku parser.

@ab5tract - `maybe wrapping or installing as a term would work`. Please can you share some sample code so that I am clear what you mean? I would be happy to try this in the interests of pinning down and fixing the bug. [But, I think that installing as a sub to be a common / the preferred way to install this kind of thing, so I would rather recommend the no-ws workaround than somehow degrade or encumber this installation method.]

@jubilatious1 - thanks for the advice!

**Comment by ab5tract** [comment]
Here's a minimum example that triggers this error:

    sub c() { 55 }
    say c / 11
    # OUTPUT: «===SORRY!===␤Regex not terminated.␤at <tmp>:1␤------> sub c { 55 }; say c / 11<HERE><EOL>␤Unable to parse regex; couldn't find final '/'␤at <tmp>:1␤------> sub c { 55 }; say c / 11<HERE><EOL>␤    expecting any of:␤        argument list…»

Here's what happens when c is defined as a term:

    sub term:<c> { 55 }
    say c / 11
    # 5

Also works like this:

    OUR::{ ‘&term:<c>’ } := sub term:<c> { 55 }
    say c / 11
    # 5

Basically term-named subs are there for the case where you never want the parser to open the question of whether said sub should take an argument.

In that case, I think this is an ideal sub form for mathematical constants.

That said, I agree that this ticket should be left open and investigated. But it could actually be that the previously working behavior changed as the result of a subtle bug _being fixed_, as opposed to one being introduced.

**Comment by jubilatious1** [comment]
@librasteve 
> I plan to evolve Physics::Measure to a RakuAST slang

I've admired the R-programming language for a while now, in particular the language's ability to name individual elements of an array:

https://stackoverflow.com/questions/66926663/is-there-a-convenient-way-to-replicate-rs-concept-of-named-vectors-in-raku-p

The examples I show in that SO post names the first element of vector `x` as `A` and gives it a value of `1`, names the second element of vector `x` as `B` and gives it a value of `2`, etc.  Further down you'll see values (areas) with island names attached, from an R built-in dataset.

Maybe of interest? I thought Mixins could replicate in Raku, but please peruse the thoughtful answers posted (and/or add your own thoughtful answer!).

TL;DR: if you keep going down the road toward providing units (postfix) for Raku values, remember that there are users interested in names (prefix) for Raku values as well!

**Comment by lizmat** [comment]
I maybe missing something, but couldn't you use `enums` for that?
```raku
my @islands = 11506,5500,16988,2968,16;
enum Places <Africa Antarctica Asia Australia>;
say @islands[Africa];  # 11506
```

**Comment by ab5tract** [comment]
Excellent point @lizmat! You can also associate values directly to enum entries:

    enum Maths (:c(55)); 
    say c / 11;
    # 5

And those values can be a Measure objects or whatever you might like (as long as you stick to a single type per enum).

Also, I believe you that this has been working with a space before but I haven't managed to reproduce a break in behavior via the bisectable bot.

Which only makes the ticket even more intriguing, of course!

**Comment by jubilatious1** [comment]
@lizmat  wrote:
> I maybe missing something, but couldn't you use `enums` for that?

Very nice answer, but Raku `enums` don't work for data with duplicated names. The built-in datasets in `R` are quite refined so you have unique entries such as `islands` and associated areas, or cities and associated `precip` (precipitation). Duplicate a dataset and you'll see `R` can still handle it:

```r
# in the R-Console (i.e. REPL)

> length(precip)
[1] 70
> two_precip <- c(precip,precip)
> length(two_precip)
[1] 140
> # find value by name
> two_precip[names(two_precip) == "Atlantic City"]
Atlantic City Atlantic City 
         45.5          45.5 
> # find value by value property
> min(two_precip)
[1] 7
> # find name/index for particular value
> which(two_precip == 45.5)
Atlantic City Atlantic City 
           38           108 
> subset by particular value
> two_precip[which(two_precip == 45.5)]
Atlantic City Atlantic City 
         45.5          45.5 
> # find value by property
> min(two_precip)
[1] 7
> # find name/index by property
> which(two_precip == min(precip))
Phoenix Phoenix 
      3      73 
> # subset vector by property to get new name/value vector
> two_precip[ which(two_precip == min(precip)) ]
Phoenix Phoenix 
      7       7 
```
________

REFERENCE DATASET (`precip`)
```r
> precip 

             Mobile              Juneau             Phoenix         Little Rock         Los Angeles 
               67.0                54.7                 7.0                48.5                14.0 
         Sacramento       San Francisco              Denver            Hartford          Wilmington 
               17.2                20.7                13.0                43.4                40.2 
         Washington        Jacksonville               Miami             Atlanta            Honolulu 
               38.9                54.5                59.8                48.3                22.9 
              Boise             Chicago              Peoria        Indianapolis          Des Moines 
               11.5                34.4                35.1                38.7                30.8 
            Wichita          Louisville         New Orleans            Portland           Baltimore 
               30.6                43.1                56.8                40.8                41.8 
             Boston             Detroit    Sault Ste. Marie              Duluth Minneapolis/St Paul 
               42.5                31.0                31.7                30.2                25.9 
            Jackson         Kansas City            St Louis         Great Falls               Omaha 
               49.2                37.0                35.9                15.0                30.2 
               Reno             Concord       Atlantic City         Albuquerque              Albany 
                7.2                36.2                45.5                 7.8                33.4 
            Buffalo            New York           Charlotte             Raleigh             Bismark 
               36.1                40.2                42.7                42.5                16.2 
         Cincinnati           Cleveland            Columbus       Oklahoma City            Portland 
               39.0                35.0                37.0                31.4                37.6 
       Philadelphia           Pittsburg          Providence            Columbia         Sioux Falls 
               39.9                36.2                42.8                46.4                24.7 
            Memphis           Nashville              Dallas             El Paso             Houston 
               49.1                46.0                35.9                 7.8                48.2 
     Salt Lake City          Burlington             Norfolk            Richmond      Seattle Tacoma 
               15.2                32.5                44.7                42.6                38.8 
            Spokane          Charleston           Milwaukee            Cheyenne            San Juan 
               17.4                40.8                29.1                14.6                59.2 
```
https://stackoverflow.com/questions/10499429/indexing-redundantly-named-vector-in-r    
https://stackoverflow.com/questions/22856926/get-named-vector-values-with-non-unique-names
